osf_tests/test_guid.py

```diff
@@ -452,3 +454,147 @@ def test_resolve_guid_download_errors(self):
 
         res = self.app.get(pp.url + 'download', auth=non_contrib.auth)
         assert res.status_code == 410
+
+
+@pytest.mark.django_db
+class TestGuidVersionsThrough:
+    def test_creation_with_guid_and_referent(self):
+        preprint = PreprintFactory()
+        guid = Guid.objects.create(referent=preprint)
+
+        version_entry = GuidVersionsThrough.objects.create(
+            guid=guid,
+            referent=preprint,
+            content_type=ContentType.objects.get_for_model(preprint),
+            object_id=preprint.pk,
+            version=1
+        )
+
+        assert version_entry.guid == guid
+        assert version_entry.referent == preprint
+        assert version_entry.content_type.model == 'preprint'
+        assert version_entry.object_id == preprint.pk
+        assert version_entry.version == 1
+        assert version_entry.is_rejected is False
+
+    def test_auto_increment_version(self):
+        preprint = PreprintFactory()
+        guid = Guid.objects.create(referent=preprint)
+
+        GuidVersionsThrough.objects.create(
+            guid=guid,
+            referent=preprint,
+            content_type=ContentType.objects.get_for_model(preprint),
+            object_id=preprint.pk,
+            version=1
+        )
+        version_entry_2 = GuidVersionsThrough.objects.create(
+            guid=guid,
+            referent=preprint,
+            content_type=ContentType.objects.get_for_model(preprint),
+            object_id=preprint.pk,
+            version=2
+        )
+
+        assert version_entry_2.version == 2
+
+    def test_creation_without_version(self):
+        preprint = PreprintFactory()
+        guid = Guid.objects.create(referent=preprint)
+
+        version_entry = GuidVersionsThrough.objects.create(
+            guid=guid,
+            referent=preprint,
+            content_type=ContentType.objects.get_for_model(preprint),
+            object_id=preprint.pk
+        )
+
+        assert version_entry.version is None
+
+    def test_referent_can_be_changed(self):
+        preprint1 = PreprintFactory()
+        preprint2 = PreprintFactory()
+        guid = Guid.objects.create(referent=preprint1)
+
+        version_entry = GuidVersionsThrough.objects.create(
+            guid=guid,
+            referent=preprint1,
+            content_type=ContentType.objects.get_for_model(preprint1),
+            object_id=preprint1.id,
+            version=1
+        )
+
+        version_entry.referent = preprint2
+        version_entry.content_type = ContentType.objects.get_for_model(preprint2)
+        version_entry.object_id = preprint2.id
+        version_entry.save()
+
+        assert version_entry.referent == preprint2
+        assert version_entry.content_type.model == 'preprint'
+        assert version_entry.object_id == preprint2.id
+
+    def test_is_rejected_flag(self):
```

Comment:
Note: we should have existing preprints machine state tests. As for the rejection case, let's see if we can add a new check for the rejected flag there. However, this is out of the scope of this ticket.